README.md
# eventemitter

[![CI](https://github.com/sagiegurari/c_eventemitter/workflows/CI/badge.svg?branch=master)](https://github.com/sagiegurari/c_eventemitter/actions)
[![Release](https://img.shields.io/github/v/release/sagiegurari/c_eventemitter)](https://github.com/sagiegurari/c_eventemitter/releases)
[![license](https://img.shields.io/github/license/sagiegurari/c_eventemitter)](https://github.com/sagiegurari/c_eventemitter/blob/master/LICENSE)

> Simple Event Emitter C API

* [Overview](#overview)
* [Usage](#usage)
* [Contributing](.github/CONTRIBUTING.md)
* [Release History](CHANGELOG.md)
* [License](#license)

<a name="overview"></a>
## Overview
This library enables to register event listeners and emit events.

<a name="usage"></a>
## Usage

```c
#include "eventemitter.h"
#include <stdio.h>

enum ExampleEvents
{
  EVENT_START  = 100,
  EVENT_MIDDLE = 200,
  EVENT_END    = 300
};


static void on_start(void *event_data, void *context)
{
  printf("OnStart, data: %s context: %s\n", (char *)event_data, (char *)context);
}


static void on_end(void *event_data, void *context)
{
  printf("OnEnd, data: %s context: %s\n", (char *)event_data, (char *)context);
}


static void on_unhandled(int event_id, void *event_data, void *context)
{
  printf("OnUnhandled, did not set listener for event: %d data: %s context: %s\n", event_id, (char *)event_data, (char *)context);
}


int main()
{
  // first create a new event emitter which will host the listeners and emit the events
  struct EventEmitter *event_emitter = eventemitter_new();

  // lets add few listeners.
  // we can add as many as we want
  // this listener will be triggered for the 'start' event
  // we are adding the same callback multiple times so for each event triggered
  // the callback will be called several times
  // The last param is a context that the specific listener will get each time its triggered
  // Each add function returns a unique ID which can be used to remove the listener
  unsigned int listener_id = eventemitter_add_listener(event_emitter, EVENT_START, on_start, "some context for start 1");

  eventemitter_add_listener(event_emitter, EVENT_START, on_start, "some context for start 2");
  eventemitter_on(event_emitter, EVENT_START, on_start, "some context for shorthand start"); // shorthand version

  // we can add listeners that are triggered once and are removed afterwards
  eventemitter_add_once_listener(event_emitter, EVENT_END, on_end, "end context 1");
  eventemitter_once(event_emitter, EVENT_END, on_end, "end context 2"); // shorthand

  // we can add listeners to any event triggered but no listeners were registered to it
  eventemitter_add_unhandled_listener(event_emitter, on_unhandled, "unhandled context 1");
  eventemitter_else(event_emitter, on_unhandled, "unhandled context 2"); // shorthand

  // trigger the 'start' event and pass some optional event data (any type) which
  // listeners will get in the callback
  eventemitter_emit(event_emitter, EVENT_START, "my start event data");

  // lets remove one of the 'start' listeners and emit it again
  eventemitter_remove_listener(event_emitter, EVENT_START, listener_id);
  eventemitter_emit(event_emitter, EVENT_START, "my second start event data");

  // lets trigger an event that no one listens to and see how the unhandled listeners are triggered
  eventemitter_emit(event_emitter, EVENT_MIDDLE, "middle event data");

  // lets trigger the 'end' event
  eventemitter_emit(event_emitter, EVENT_END, "end event data");
  // since all 'end' event listeners were added as 'once' they are no longer registered
  // so we will trigger it again and see how the unhandled listeners are invoked
  eventemitter_emit(event_emitter, EVENT_END, "end event data again");
  eventemitter_emit(event_emitter, EVENT_END, "end event data again 2");

  // once done with the emitter, we should release it
  // no need to release the listeners manually
  printf("Releasing Event Emitter\n");
  eventemitter_release(event_emitter);

  return(0);
} /* main */
```

## Contributing
See [contributing guide](.github/CONTRIBUTING.md)

<a name="history"></a>
## Release History

See [Changelog](CHANGELOG.md)

<a name="license"></a>
## License
Developed by Sagie Gur-Ari and licensed under the Apache 2 open source license.


CHANGELOG.md
## CHANGELOG

### v0.1.1

* Added void to no arg functions
* Updated header include guard macro name

### v0.1.0 (2022-04-19)

* Initial Release


tests/test.c
#include "test.h"
#include <stdio.h>
#include <string.h>


void test_run(void (*fn)(void))
{
  printf("Test ... ");
  fn();
  printf("Done\n");
}


void test_fail()
{
  printf(" Error\n");
  exit(1);
}


void assert_true(bool value)
{
  if (!value)
  {
    test_fail();
  }
}


void assert_num_equal(size_t value1, size_t value2)
{
  if (value1 != value2)
  {
#ifdef linux
    printf("Assert Failed, value: %zu not equals to value: %zu", value1, value2);
#endif
    test_fail();
  }
}


void assert_string_equal(char *value1, char *value2)
{
  if (strcmp(value1, value2) != 0)
  {
    printf("assert failed, value: %s not equals to value: %s", value1, value2);
    test_fail();
  }
}



tests/test.h
#ifndef __TEST_H__
#define __TEST_H__

#include "eventemitter.h"
#include <stdlib.h>

void test_run(void (*fn)());
void test_fail();
void assert_true(bool);

void assert_num_equal(size_t, size_t);
void assert_string_equal(char *, char *);

#endif



tests/test_release_null.c
#include "test.h"


void test_impl()
{
  eventemitter_release(NULL);
}


int main()
{
  test_run(test_impl);
}



src/eventemitter.c
#include "eventemitter.h"
#include "vector.h"
#include <stdlib.h>

struct EventEmitter
{
  unsigned int  next_callback_id;
  struct Vector *event_listeners;
  struct Vector *unhandled_listeners;
};

struct EventEmitterEventListeners
{
  int           event_id;
  struct Vector *listeners;
};

struct EventEmitterEventListener
{
  unsigned int id;
  void         (*callback)(void *event_data, void *context);
  void         *context;
  bool         once;
};

struct EventEmitterUnhandledListener
{
  unsigned int id;
  void         (*callback)(int event_id, void *event_data, void *context);
  void         *context;
};

// private functions
static struct EventEmitterEventListeners *_eventemitter_get_listeners_for_event_id(struct EventEmitter *, int);
static unsigned int _eventemitter_add_listener(struct EventEmitter *, int, void (*callback)(void *, void *), void *, bool, bool);
static unsigned int _eventemitter_add_unhandled_listener(struct EventEmitter *, void (*callback)(int, void *, void *), void *, bool);

struct EventEmitter *eventemitter_new()
{
  struct EventEmitter *event_emitter = malloc(sizeof(struct EventEmitter));

  event_emitter->next_callback_id    = 1;
  event_emitter->event_listeners     = vector_new();
  event_emitter->unhandled_listeners = vector_new();

  return(event_emitter);
}


void eventemitter_release(struct EventEmitter *event_emitter)
{
  if (event_emitter == NULL)
  {
    return;
  }

  eventemitter_remove_all_listeners(event_emitter);
  vector_release(event_emitter->event_listeners);
  vector_release(event_emitter->unhandled_listeners);
  free(event_emitter);
}


unsigned int eventemitter_add_listener(struct EventEmitter *event_emitter, int event_id, void (*callback)(void *event_data, void *context), void *context)
{
  return(_eventemitter_add_listener(event_emitter, event_id, callback, context, false, false));
}


unsigned int eventemitter_on(struct EventEmitter *event_emitter, int event_id, void (*callback)(void *event_data, void *context), void *context)
{
  return(_eventemitter_add_listener(event_emitter, event_id, callback, context, false, false));
}


unsigned int eventemitter_prepend_listener(struct EventEmitter *event_emitter, int event_id, void (*callback)(void *event_data, void *context), void *context)
{
  return(_eventemitter_add_listener(event_emitter, event_id, callback, context, false, true));
}


unsigned int eventemitter_add_once_listener(struct EventEmitter *event_emitter, int event_id, void (*callback)(void *event_data, void *context), void *context)
{
  return(_eventemitter_add_listener(event_emitter, event_id, callback, context, true, false));
}


unsigned int eventemitter_once(struct EventEmitter *event_emitter, int event_id, void (*callback)(void *event_data, void *context), void *context)
{
  return(_eventemitter_add_listener(event_emitter, event_id, callback, context, true, false));
}


unsigned int eventemitter_prepend_once_listener(struct EventEmitter *event_emitter, int event_id, void (*callback)(void *event_data, void *context), void *context)
{
  return(_eventemitter_add_listener(event_emitter, event_id, callback, context, true, true));
}


unsigned int eventemitter_add_unhandled_listener(struct EventEmitter *event_emitter, void (*callback)(int event_id, void *event_data, void *context), void *context)
{
  return(_eventemitter_add_unhandled_listener(event_emitter, callback, context, false));
}


unsigned int eventemitter_else(struct EventEmitter *event_emitter, void (*callback)(int event_id, void *event_data, void *context), void *context)
{
  return(_eventemitter_add_unhandled_listener(event_emitter, callback, context, false));
}


unsigned int eventemitter_prepend_unhandled_listener(struct EventEmitter *event_emitter, void (*callback)(int event_id, void *event_data, void *context), void *context)
{
  return(_eventemitter_add_unhandled_listener(event_emitter, callback, context, true));
}


int eventemitter_remove_listener(struct EventEmitter *event_emitter, int event_id, unsigned int callback_id)
{
  if (event_emitter == NULL || !callback_id)
  {
    return(-1);
  }

  struct EventEmitterEventListeners *listeners = _eventemitter_get_listeners_for_event_id(event_emitter, event_id);
  if (listeners == NULL)
  {
    return(0);
  }

  size_t count  = vector_size(listeners->listeners);
  int    output = 0;
  for (size_t index = 0; index < count; index++)
  {
    struct EventEmitterEventListener *listener = (struct EventEmitterEventListener *)vector_get(listeners->listeners, index);

    if (listener != NULL && listener->id == callback_id)
    {
      free(listener);
      vector_remove(listeners->listeners, index);
      output = 1;
      break;
    }
  }

  if (vector_is_empty(listeners->listeners))
  {
    eventemitter_remove_all_event_listeners(event_emitter, event_id);
  }

  return(output);
}


int eventemitter_off(struct EventEmitter *event_emitter, int event_id, unsigned int callback_id)
{
  return(eventemitter_remove_listener(event_emitter, event_id, callback_id));
}


int eventemitter_remove_unhandled_listener(struct EventEmitter *event_emitter, unsigned int callback_id)
{
  if (event_emitter == NULL || !callback_id)
  {
    return(-1);
  }

  size_t count  = vector_size(event_emitter->unhandled_listeners);
  int    output = 0;
  for (size_t index = 0; index < count; index++)
  {
    struct EventEmitterUnhandledListener *listener = (struct EventEmitterUnhandledListener *)vector_get(event_emitter->unhandled_listeners, index);

    if (listener != NULL && listener->id == callback_id)
    {
      free(listener);
      vector_remove(event_emitter->unhandled_listeners, index);
      output = 1;
      break;
    }
  }

  return(output);
}


bool eventemitter_remove_all_event_listeners(struct EventEmitter *event_emitter, int event_id)
{
  if (event_emitter == NULL)
  {
    return(false);
  }

  struct EventEmitterEventListeners *listeners = _eventemitter_get_listeners_for_event_id(event_emitter, event_id);
  if (listeners == NULL)
  {
    return(true);
  }

  size_t count = vector_size(listeners->listeners);
  for (size_t index = 0; index < count; index++)
  {
    struct EventEmitterEventListener *listener = (struct EventEmitterEventListener *)vector_get(listeners->listeners, index);

    if (listener != NULL)
    {
      free(listener);
    }
  }

  count = vector_size(event_emitter->event_listeners);
  for (size_t index = 0; index < count; index++)
  {
    listeners = (struct EventEmitterEventListeners *)vector_get(event_emitter->event_listeners, index);

    if (listeners != NULL && listeners->event_id == event_id)
    {
      vector_remove(event_emitter->event_listeners, index);
      vector_release(listeners->listeners);
      free(listeners);
    }
  }

  return(true);
} /* eventemitter_remove_all_event_listeners */


bool eventemitter_remove_all_unhandled_listeners(struct EventEmitter *event_emitter)
{
  if (event_emitter == NULL)
  {
    return(false);
  }

  size_t count = vector_size(event_emitter->unhandled_listeners);
  for (size_t index = 0; index < count; index++)
  {
    struct EventEmitterUnhandledListener *listener = (struct EventEmitterUnhandledListener *)vector_get(event_emitter->unhandled_listeners, index);

    if (listener != NULL)
    {
      free(listener);
    }
  }
  vector_clear(event_emitter->unhandled_listeners);

  return(true);
}


bool eventemitter_remove_all_listeners(struct EventEmitter *event_emitter)
{
  if (event_emitter == NULL)
  {
    return(false);
  }

  while (!vector_is_empty(event_emitter->event_listeners))
  {
    struct EventEmitterEventListeners *listeners = (struct EventEmitterEventListeners *)vector_get(event_emitter->event_listeners, 0);

    if (listeners != NULL)
    {
      eventemitter_remove_all_event_listeners(event_emitter, listeners->event_id);
    }
    else
    {
      vector_remove(event_emitter->event_listeners, 0);
    }
  }

  eventemitter_remove_all_unhandled_listeners(event_emitter);

  return(true);
}


int eventemitter_listeners_count(struct EventEmitter *event_emitter, int event_id)
{
  if (event_emitter == NULL)
  {
    return(-1);
  }

  struct EventEmitterEventListeners *listeners = _eventemitter_get_listeners_for_event_id(event_emitter, event_id);
  if (listeners == NULL)
  {
    return(0);
  }

  return((int)vector_size(listeners->listeners));
}


int eventemitter_emit(struct EventEmitter *event_emitter, int event_id, void *event_data)
{
  if (event_emitter == NULL)
  {
    return(-1);
  }

  int                               callback_counter = 0;
  struct EventEmitterEventListeners *listeners       = _eventemitter_get_listeners_for_event_id(event_emitter, event_id);
  if (listeners != NULL)
  {
    size_t count = vector_size(listeners->listeners);
    for (size_t index = 0; index < count; index++)
    {
      struct EventEmitterEventListener *listener = (struct EventEmitterEventListener *)vector_get(listeners->listeners, index);

      if (listener != NULL)
      {
        listener->callback(event_data, listener->context);
        callback_counter++;
      }
    }

    // remove 'once' listeners
    for (size_t index = 0; index < count; index++)
    {
      size_t                           end_index = count - 1 - index;
      struct EventEmitterEventListener *listener = (struct EventEmitterEventListener *)vector_get(listeners->listeners, end_index);

      if (listener != NULL && listener->once)
      {
        free(listener);
        vector_remove(listeners->listeners, end_index);
      }
    }
    if (vector_is_empty(listeners->listeners))
    {
      eventemitter_remove_all_event_listeners(event_emitter, event_id);
    }
  }
  else
  {
    size_t count = vector_size(event_emitter->unhandled_listeners);
    for (size_t index = 0; index < count; index++)
    {
      struct EventEmitterUnhandledListener *listener = (struct EventEmitterUnhandledListener *)vector_get(event_emitter->unhandled_listeners, index);

      if (listener != NULL)
      {
        listener->callback(event_id, event_data, listener->context);
        callback_counter++;
      }
    }
  }

  return(callback_counter);
} /* eventemitter_emit */

static struct EventEmitterEventListeners *_eventemitter_get_listeners_for_event_id(struct EventEmitter *event_emitter, int event_id)
{
  if (event_emitter == NULL)
  {
    return(NULL);
  }

  size_t count = vector_size(event_emitter->event_listeners);
  for (size_t index = 0; index < count; index++)
  {
    struct EventEmitterEventListeners *listeners = (struct EventEmitterEventListeners *)vector_get(event_emitter->event_listeners, index);

    if (listeners != NULL && listeners->event_id == event_id)
    {
      return(listeners);
    }
  }

  return(NULL);
}


static unsigned int _eventemitter_add_listener(struct EventEmitter *event_emitter, int event_id, void (*callback)(void *event_data, void *context), void *context, bool once, bool prepend)
{
  if (event_emitter == NULL || callback == NULL)
  {
    return(0);
  }

  struct EventEmitterEventListeners *listeners = _eventemitter_get_listeners_for_event_id(event_emitter, event_id);
  if (listeners == NULL)
  {
    listeners            = malloc(sizeof(struct EventEmitterEventListeners));
    listeners->event_id  = event_id;
    listeners->listeners = vector_new();
    vector_push(event_emitter->event_listeners, listeners);
  }

  // allocate next id for listener
  unsigned int callback_id = event_emitter->next_callback_id;
  event_emitter->next_callback_id++;

  // create listener wrapper
  struct EventEmitterEventListener *listener = malloc(sizeof(struct EventEmitterEventListener));
  listener->id       = callback_id;
  listener->callback = callback;
  listener->context  = context;
  listener->once     = once;

  // keep in event listeners list
  if (prepend)
  {
    vector_prepend(listeners->listeners, listener);
  }
  else
  {
    vector_push(listeners->listeners, listener);
  }

  return(callback_id);
}


static unsigned int _eventemitter_add_unhandled_listener(struct EventEmitter *event_emitter, void (*callback)(int, void *, void *), void *context, bool prepend)
{
  if (event_emitter == NULL || callback == NULL)
  {
    return(0);
  }

  // allocate next id for listener
  unsigned int callback_id = event_emitter->next_callback_id;
  event_emitter->next_callback_id++;

  // create listener wrapper
  struct EventEmitterUnhandledListener *listener = malloc(sizeof(struct EventEmitterUnhandledListener));
  listener->id       = callback_id;
  listener->callback = callback;
  listener->context  = context;

  // keep in event listeners list
  if (prepend)
  {
    vector_prepend(event_emitter->unhandled_listeners, listener);
  }
  else
  {
    vector_push(event_emitter->unhandled_listeners, listener);
  }

  return(callback_id);
}



include/eventemitter.h
#ifndef EVENTEMITTER_H
#define EVENTEMITTER_H

#include <stdbool.h>

struct EventEmitter;

/**
 * Creates and returns a new event emitter.
 * Once no longer needed, it must be released.
 */
struct EventEmitter *eventemitter_new(void);

/**
 * Frees the memory of the provided emitter.
 */
void eventemitter_release(struct EventEmitter *);

/**
 * Adds a new event listener for the given event ID.
 * The provided optional context will be provided to the callback when the event is triggered.
 * The returned ID can be later used to remove the listener.
 * In case of invalid input, 0 will be returned.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_add_listener(struct EventEmitter *, int /* event ID */, void (*callback)(void * /* event data */, void * /* context */), void * /* context */);

/**
 * Shorthand for add_listener.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_on(struct EventEmitter *, int /* event ID */, void (*callback)(void * /* event data */, void * /* context */), void * /* context */);

/**
 * Same as the add listener, but it adds it to the start of the listener list.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_prepend_listener(struct EventEmitter *, int /* event ID */, void (*callback)(void * /* event data */, void * /* context */), void * /* context */);

/**
 * Same as the add listener, but once an event is handled by the callback,
 * the callback would automatically be removed.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_add_once_listener(struct EventEmitter *, int /* event ID */, void (*callback)(void * /* event data */, void * /* context */), void * /* context */);

/**
 * Shorthand for add once listener.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_once(struct EventEmitter *, int /* event ID */, void (*callback)(void * /* event data */, void * /* context */), void * /* context */);

/**
 * Same as the add once listener, but it adds it to the start of the listener list.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_prepend_once_listener(struct EventEmitter *, int /* event ID */, void (*callback)(void * /* event data */, void * /* context */), void * /* context */);

/**
 * Adds a new event listener which will be invoked for any event ID without registered listeners.
 * The provided optional context will be provided to the callback when the event is triggered.
 * The returned ID can be later used to remove the listener.
 * In case of invalid input, 0 will be returned.
 *
 * @param event emitter - The emitter struct
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_add_unhandled_listener(struct EventEmitter *, void (*callback)(int /* event ID */, void * /* event data */, void * /* context */), void * /* context */);

/**
 * Shorthand for add unhandled listener.
 *
 * @param event emitter - The emitter struct
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_else(struct EventEmitter *, void (*callback)(int /* event ID */, void * /* event data */, void * /* context */), void * /* context */);

/**
 * Same as the add unhandled listener, but it adds it to the start of the listener list.
 *
 * @param event emitter - The emitter struct
 * @param callback - Will be called when the event ID is triggered via emit
 * @param context - Will be passed to this specific callback when an event is triggered
 * @returns 0 in case of error or the callback ID which can be used to remove the listener
 */
unsigned int eventemitter_prepend_unhandled_listener(struct EventEmitter *, void (*callback)(int /* event ID */, void * /* event data */, void * /* context */), void * /* context */);

/**
 * Removes the listener for the given event ID and callback ID if exists.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback ID - The callback ID returned from the add listener function
 * @returns -1 for invalid input, 0 for callback not found, 1 for removed
 */
int eventemitter_remove_listener(struct EventEmitter *, int /* event ID */, unsigned int /* callback ID */);

/**
 * Shorthand for remove listener.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @param callback ID - The callback ID returned from the add listener function
 * @returns -1 for invalid input, 0 for callback not found, 1 for removed
 */
int eventemitter_off(struct EventEmitter *, int /* event ID */, unsigned int /* callback ID */);

/**
 * Removes the unhandled events listener for the given callback ID if exists.
 *
 * @param event emitter - The emitter struct
 * @param callback ID - The callback ID returned from the add listener function
 * @returns -1 for invalid input, 0 for callback not found, 1 for removed
 */
int eventemitter_remove_unhandled_listener(struct EventEmitter *, unsigned int /* callback ID */);

/**
 * Removes all listeners for the given event ID.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @returns true in case of valid input
 */
bool eventemitter_remove_all_event_listeners(struct EventEmitter *, int /* event ID */);

/**
 * Removes all unhandled event listeners.
 *
 * @param event emitter - The emitter struct
 * @returns true in case of valid input
 */
bool eventemitter_remove_all_unhandled_listeners(struct EventEmitter *);

/**
 * Removes all listeners for all events.
 *
 * @param event emitter - The emitter struct
 * @returns true in case of valid input
 */
bool eventemitter_remove_all_listeners(struct EventEmitter *);

/**
 * Returns the listeners count for the given event ID.
 * In case of invalid input, -1 will be returned.
 *
 * @param event emitter - The emitter struct
 * @param event ID - The event ID that listeners have registered on
 * @returns the amount of callbacks registered for the given event ID or returns -1 in case of invalid input
 */
int eventemitter_listeners_count(struct EventEmitter *, int /* event ID */);

/**
 * Triggers an event for the given event ID.
 * The event data would be given to all relevant event listeners.
 *
 * @param event emitter - The emitter struct
 * @param event data - The event data passed to all relevant listeners
 * @returns the amount of callbacks invoked (including unhandled) or returns -1 in case of invalid input
 */
int eventemitter_emit(struct EventEmitter *, int /* event ID */, void * /* event data */);

#endif



examples/example.c
#include "eventemitter.h"
#include <stdio.h>

enum ExampleEvents
{
  EVENT_START  = 100,
  EVENT_MIDDLE = 200,
  EVENT_END    = 300
};


static void on_start(void *event_data, void *context)
{
  printf("OnStart, data: %s context: %s\n", (char *)event_data, (char *)context);
}


static void on_end(void *event_data, void *context)
{
  printf("OnEnd, data: %s context: %s\n", (char *)event_data, (char *)context);
}


static void on_unhandled(int event_id, void *event_data, void *context)
{
  printf("OnUnhandled, did not set listener for event: %d data: %s context: %s\n", event_id, (char *)event_data, (char *)context);
}


int main()
{
  // first create a new event emitter which will host the listeners and emit the events
  struct EventEmitter *event_emitter = eventemitter_new();

  // lets add few listeners.
  // we can add as many as we want
  // this listener will be triggered for the 'start' event
  // we are adding the same callback multiple times so for each event triggered
  // the callback will be called several times
  // The last param is a context that the specific listener will get each time its triggered
  // Each add function returns a unique ID which can be used to remove the listener
  unsigned int listener_id = eventemitter_add_listener(event_emitter, EVENT_START, on_start, "some context for start 1");

  eventemitter_add_listener(event_emitter, EVENT_START, on_start, "some context for start 2");
  eventemitter_on(event_emitter, EVENT_START, on_start, "some context for shorthand start"); // shorthand version

  // we can add listeners that are triggered once and are removed afterwards
  eventemitter_add_once_listener(event_emitter, EVENT_END, on_end, "end context 1");
  eventemitter_once(event_emitter, EVENT_END, on_end, "end context 2"); // shorthand

  // we can add listeners to any event triggered but no listeners were registered to it
  eventemitter_add_unhandled_listener(event_emitter, on_unhandled, "unhandled context 1");
  eventemitter_else(event_emitter, on_unhandled, "unhandled context 2"); // shorthand

  // trigger the 'start' event and pass some optional event data (any type) which
  // listeners will get in the callback
  eventemitter_emit(event_emitter, EVENT_START, "my start event data");

  // lets remove one of the 'start' listeners and emit it again
  eventemitter_remove_listener(event_emitter, EVENT_START, listener_id);
  eventemitter_emit(event_emitter, EVENT_START, "my second start event data");

  // lets trigger an event that no one listens to and see how the unhandled listeners are triggered
  eventemitter_emit(event_emitter, EVENT_MIDDLE, "middle event data");

  // lets trigger the 'end' event
  eventemitter_emit(event_emitter, EVENT_END, "end event data");
  // since all 'end' event listeners were added as 'once' they are no longer registered
  // so we will trigger it again and see how the unhandled listeners are invoked
  eventemitter_emit(event_emitter, EVENT_END, "end event data again");
  eventemitter_emit(event_emitter, EVENT_END, "end event data again 2");

  // once done with the emitter, we should release it
  // no need to release the listeners manually
  printf("Releasing Event Emitter\n");
  eventemitter_release(event_emitter);

  return(0);
} /* main */
